### ISSUE TITLE
Applescript Installation Notification

### ISSUE THREAD

**Issue by shawnrice** [shawnrice/alfred-bundler#49]
So, I was thinking about adding an Applescript dialog that tells you when you're about to install the bundler and asks permission.

For a quick check, delete the installed bundler directory, and then just run the PHP bundler from the command line.

Tell me what you think. Good idea? Annoying?

---
#### Tasks
- [x] Solidify dialog text
- [x] Choose icon
- [x] Decide behavior when user denies installation
- [x] Decide behavior when user presses `more info` button
- [x] Implement into Bash
- [x] Implement into PHP
- [x] Implement into Python
- [x] Implement into Ruby

**Comment by deanishe** [comment]
I like it. Always a good idea to let the user know before you install stuff and inform them why their workflow is taking so long.

**Comment by shawnrice** [comment]
Sweet. The text in the dialog box can easily be updated (consider it placeholder text), and the icon can also be changed. We can actually use one of the system icons to make it a bit nicer.

In the PHP version, if the user chooses not to install the bundler, then it reports it in STDERR and exits the bundler script. But it doesn't kill the top process. Also, I changed the PHP bundler so that (theoretically) every call to the bundler that cannot be completed reports to STDERR but lets the workflow keep operating however well it can without the bundler.

Lastly, we can change the page that the "more info" button links to. I just wrote up a detailed explanation [on the wiki](https://github.com/shawnrice/alfred-bundler/wiki/What-is-the-Alfred-Bundler) that might be a good place to point.

So, let's make this a task. I think it should be implemented natively in each wrapper before the install script is either downloaded or called.

**Comment by deanishe** [comment]
With regard to errors, in the Python version I raise Exceptions whenever something goes wrong. That's the common thing to do, and it's up to the user to catch (and check) the errors and decide whether to proceed or not.

**Comment by shawnrice** [comment]
Most people writing workflows with PHP don't use the `try... catch...` to check exceptions, so I'm just returning false to cater to how most PHP workflows that I've seen are written.

If it's common in Python to do so, then that is obviously the way to go.

I wonder if it's common in Ruby...

**Comment by deanishe** [comment]
Yeah, PHP doesn't have a great culture when it comes to error handling…

I honestly don't know regarding Ruby, but I hope it's closer to the Python way of having to handle errors by default instead of PHP's ignoring errors by default.

Shit, they even have a mantra for this in the Python community: "Easier to ask forgiveness than permission".

**Comment by shawnrice** [comment]
Well, I'm assuming that our implementation of error handling in Ruby will start off in not the most amazing state and then get better. Yay for a self-updating framework.

**Comment by shawnrice** [comment]
Let's decide on an icon:

Here are some that are native in each operating system:
- ToolbarInfo ![toolbarinfo](https://cloud.githubusercontent.com/assets/3452150/3867625/7fe1d72c-2008-11e4-93b7-11625827158e.png)
- SideBarDownloadsFolder ![sidebardownloadsfolder](https://cloud.githubusercontent.com/assets/3452150/3867624/7fdf114a-2008-11e4-885c-56e9b3f65b70.png)
- GenericQuestionMarkIcon ![genericquestionmarkicon](https://cloud.githubusercontent.com/assets/3452150/3867626/7fe1fab8-2008-11e4-965e-683465d73b03.png)
- AlertNoteIcon ![alertnoteicon](https://cloud.githubusercontent.com/assets/3452150/3867627/7fe7391a-2008-11e4-9fa8-546240121509.png)

I don't want to use the Alfred icon.

**Comment by shawnrice** [comment]
Let's have the more info button go to the [wiki explanation page](https://github.com/shawnrice/alfred-bundler/wiki/What-is-the-Alfred-Bundler). It's easy enough to keep updated, and we will have links there to the `gh-pages` documentation as well.

**Comment by shawnrice** [comment]
Let's edit the message here.

> {%Workflow Name%} has requested to install the Alfred Bundler, a framework written for Alfred2 to provide more features for workflows.
> 
> Don't worry, the Bundler and all data that it downloads stay in Alfred's data directory, away from your system.
> 
> There will be a slight delay after accepting. To learn more about what the bundler is and how it works, click the 'more information' button.

**Comment by shawnrice** [comment]
Below is the current Applescript code for the dialog, reflecting any decisions made above.

> display dialog "{%Workflow Name%} has requested to install the Alfred Bundler, a framework written for Alfred2 to provide more features for workflows.
> 
> Don't worry, the Bundler and all data that it downloads stay in Alfred's data directory, away from your system.
> 
> There will be a slight delay after accepting. To learn more about what the bundler is and how it works, click the 'more information' button." buttons {"More Info", "Abort", "Proceed"} default button 3 with title "Install Alfred Bundler" with icon file "System:Library:CoreServices:CoreTypes.bundle:Contents:Resources:SideBarDownloadsFolder.icns"

![screen shot 2014-08-09 at 5 27 59 pm](https://cloud.githubusercontent.com/assets/3452150/3867715/142580b6-200c-11e4-878d-ce3d23367aef.PNG)

**Comment by stephen-bunn** [comment]
Just to clarify, button 3 is the "Proceed" button?

**Comment by shawnrice** [comment]
Yes.

**Comment by stephen-bunn** [comment]
Looks good to me, although shouldn't we or I create an AppleScriptDialog wrapper similar to the current wrappers, rather than hard coding the AppleScript in?

**Comment by shawnrice** [comment]
Actually, these are now in the bundlets, so we can't use any wrappers. Since the workflow author will include the bundlet, this AS will be shown before the installation is called within the bundlet, so we can't use any external assets, and thus it needs to be hardcoded.

**Comment by shawnrice** [comment]
Oh, last element: Title.
1. Install Alfred Bundler
2. {%Workflow Name%} Setup

The latter is valid because using the bundler means that setting up the workflow includes bundler installation. Also, it seems less scary because it is identified.

**Comment by stephen-bunn** [comment]
It would probably be better to use the 2nd since it sounds less intimidating than "Install Asset Bundler"

**Comment by deanishe** [comment]
My ten euro cents:

I think we should either use the bundler icon or the workflow icon.

I think the text is too long to expect people to read. I think we need to mention four things:
1. Workflow XYZ needs to install some extras
2. You may have to give permission for some of these extras to run
3. If you refuse either, the workflow won't work
4. None of the extras can possibly interfere with the system
5. This message brought to you by Alfred Dependency Bundler (see comment below)

So as a first draft:

> {%Workflow Name%} needs to install additional components. These will be installed in Alfred's storage and will not interfere with your system in any way.
> 
> You may be asked to allow some components to run, depending on your Gatekeeper settings.
> 
> You can refuse permission to install/run these components, but the workflow will not work without them.

That's not really any shorter, is it? :frowning: 

Any message we post is necessarily excessively vague (we don't know what the workflow actually wants to install), so we should mention the bundler in some way, perhaps in the dialog title:
- Alfred Bundler Asset Installer by Shawn Rice
- Packal Workflow Bundler by Shawn Rice
- ???

I think including Shawn's/Packal's name might allay potential worries, especially with regard to the vagueness of the message.

It might also be a good idea to allow users to override the install message (and icon?), so they can provide a more specific message.

**Comment by shawnrice** [comment]
I think that the title of `{%Workflow Name%} Setup` would allay fears even more than including my name / Packal because they just _called_ that workflow. While Packal will probably become (/is) the most recognizable name when it comes to Alfred Workflows, I don't know if my name carries the same weight. If the person reads the forums enough, then mine and yours, @deanishe, should be recognizable enough because we both have 600+ posts.

For the message, I don't know how well recognized the word `Gatekeeper` is. Maybe just "depending on your security settings" instead.

Based on conversations that I've had with Andrew, I believe we need to mention the delay.

**Comment by deanishe** [comment]
> I think that the title of {%Workflow Name%} Setup would allay fears even more than including my name

Initially, definitely. Longer term, if the bundler becomes a thing, then its (your) name will carry more weight.

> While Packal will probably become (/is) the most recognizable name when it comes to Alfred Workflows

Hence the suggestion. I suggested your name, as a person's name implies greater responsibility than that of a "corporation".

> Maybe just "depending on your security settings" instead.

You're spot-on; that's definitely better.

> Based on conversations that I've had with Andrew, I believe we need to mention the delay.

Good point. Indeed we do. Andrew's super-keen on performance, and he's right to be.

So, got a new text?

**Comment by shawnrice** [comment]
Super awkward on how to integrate my name.

**Comment by shawnrice** [comment]
New edition:

> {%Workflow Name%} needs to install additional components, which will be placed in Alfred's storage and will not interfere with your system.
> 
> You may be asked to allow some components to run, depending on your security settings.
> 
> You can decline this installation, but {%Workflow Name%} may not work without them. There will be a slight delay after accepting.

**Comment by shawnrice** [comment]
Oh, we need to define a common behavior on what to do if the user decides not to install the bundler. Do we:
1. `die()` (tempting, but probably a bad idea)
2. Skip the installation and try to recuperate any and all errors that would occur without the bundler installed?

Further, what do we do when the user presses the `more info` button:
1. `die()` (makes more sense here)
2. Wait....? (how?)

**Comment by deanishe** [comment]
If the user says no, the answer is simple. Throw an error and let the workflow author handle it. The bundler cannot possibly know what the consequences are for any given workflow.

The More Info one is tricky. I still don't think killing the process is a good idea: it's not the bundler's process to kill.

In Python, there is already an `InstallationError` exception. I will add `UserRefused` and `MoreInfo` subclasses. Authors can just catch `InstallationError` and check the exact class if they want to know more.

**Comment by shawnrice** [comment]
That's a good method, and maybe the one we should ultimately go with. The possible problem is how people (especially people who write Alfred workflows) code in these languages.

Exception handling is fairly standard practice in Python and Ruby (from what I know), but it isn't often or well implemented with PHP. If we do throw the exception with PHP, then I would need to / happily create a bit of documentation that explains how to catch any thrown exceptions. The result would be that author would write better code.

But, the Bash one is harder to simulate the same behavior. We could set some traps that would catch the errors generated by the bundler but let the other ones through (the conditional might be really hard to code). The fallback method that we could use for bash would just be to let the process kill itself via calling an unknown function.

**Comment by shawnrice** [comment]
I actually do like killing the process on the "More Info" because it seems to follow the logic of user interaction. If I press a "more info" button, I expect that whatever I was doing to be put on hold / canceled while I go read up on it. So, it kind of makes sense there.

**Comment by shawnrice** [comment]
Actually, with Bash, we could set the trap to respond only to a rarely used exit code, and then just throw a message with that and no _actually_ exit the process.

**Comment by deanishe** [comment]
I think that throwing an Exception is definitely the right way to go in PHP: it will have essentially the same effect as `die()`, but also allow authors to catch it if they want to. It forces them to code properly instead of chugging on regardless with a broken workflow.

I'm not so sure with bash. throwing exceptions and adding listeners doesn't seem to be that common, and requires a fair bit of boilerplate from the workflow author. Might it be better to give the bundler an explicit `initialize` function that a workflow has to call? That way, workflows can check the return code, which is easier and more common.

**Comment by shawnrice** [comment]
I guess that `initialize` for Bash would be the same as instantiating an object in PHP.

For Bash, further, what about this:
User denies installation. We set a flag that indicates as much. After that, we set a trap to catch every 127 error and see if the 127 error was caused by something that had a bash bundler prefix. If so, we send a message to STDERR and send a return 1 rather than a 127. And then we let all the other 127 exit codes kill the process like they normally would.

**Comment by deanishe** [comment]
I don't follow you. Why do you want to trap errors?

**Comment by shawnrice** [comment]
TL;DR: Basically, it would turn many unrecoverable errors into recoverable exceptions. 

127 is command not found. So, when a user denies installation, then almost all of the  bundler functions that the workflow would call after that would result in the script breaking. Trapping and filtering the 127's would make it so that the workflow could keep going by throwing recoverable exceptions.

So, we'd pass the 127 bundler function not found error to a skeleton function defined in the bundlet. The workflow author could then handle exceptions by overwriting the function for error handling. Or just let them die.

All other 127's would slip through our trap, having the script die as normal.

**Comment by shawnrice** [comment]
Okay. Here's an example implementation. Run the script at `{bundler_dir}/tests/bash/bash_exceptions.sh` and see it fail. I need to doublecheck to make sure it'll only die on the right codes (maybe I need to trap `EXIT` instead of `ERR`), but it shows what's happening.

**Comment by shawnrice** [comment]
Added preliminary support for the AS in Bash in 09b5c209409fc6313056b8fb920e95cbcd451d86. It hasn't been integrated yet. I didn't put in any traps. Don't worry. Yet.

**Comment by deanishe** [comment]
Wouldn't having an `initialize` function be a lot simpler than faffing about with trapping errors? If it returns 0, the bundler is good to go, if it doesn't something went wrong. The user can determine what went wrong based on the return code.

**Comment by shawnrice** [comment]
That is easier. We'll do that.

**Comment by shawnrice** [comment]
Okay, I'm putting this out as a decision, which I think is solid and follows what we've talked about before.

Decision first. Sample code second. Explanation last.

The text of the applescript dialog should be:

> "{%Workflow Name%} needs to install additional components, which will be placed in the Alfred storage directory and will not interfere with your system.
> 
> You may be asked to allow some components to run, depending on your security settings.
> 
> You can decline this installation, but {%Workflow Name%} may not work without them. There will be a slight delay after accepting."

with the title

> {%Workflow Name%} Setup

When `More Info` is pressed, we should kill the process.

When `Cancel` is pressed, we should (as best as the language can) throw an exception for the workflow author to deal with.

The icon should try to be the `icon.png` in the workflow folder and should fallback to `System:Library:CoreServices:CoreTypes.bundle:Contents:Resources:SideBarDownloadsFolder.icns`.

```
  if ( file_exists( 'icon.png' ) ) {
    $icon = realpath( 'icon.png' );
    $icon = str_replace('/', ':', 'icon');
    $icon = substr( $icon, 1, strlen( $icon ) - 1 );
  } else {
    $icon = "System:Library:CoreServices:CoreTypes.bundle:Contents:Resources:SideBarDownloadsFolder.icns";
  }
  // Text for the dialog message.
  $text = "{$this->name} needs to install additional components, which will be placed in the Alfred storage directory and will not interfere with your system.

You may be asked to allow some components to run, depending on your security settings.

You can decline this installation, but {$this->name} may not work without them. There will be a slight delay after accepting.";

  $this->script = "display dialog \"$text\" " .
    "buttons {\"More Info\",\"Cancel\",\"Proceed\"} default button 3 " .
    "with title \"{$this->name} Setup\" with icon file \"$icon\"";
```

Obvs: code is PHP. The icon path needs to be a full path, separated by semi-colons not forward slashes, and must not start with a semi-colon. The PHP code above shows it a bit more "step-by-step."

Explanations:

Title: `{%Workflow Name%} Setup` makes sense because installing the bundler is part of the workflow setup. It helps cue the user to understand that this is part of the workflow they just downloaded and ran — not something completely different. The text of the message clear communicates that it is an adjunct framework.

Killing PID on `More Info` makes sense as it follows expected behavior in two ways: pressing a button to find out more presses pause on whatever you are doing so that you can come back to it after you've gotten more information. But, it's hard to re-start a script after it has been paused (from the Alfred interface). Further, since pressing the button on the AS dialog will open a browser window and keep a script filter window from re-populating, it does kill the process.

Exception on `Cancel`: needs no explanation. But, we should set an environmental variable for the script
`$_ENV[ 'ALFRED_BUNDLER_INSTALL_REFUSED' ] = TRUE;` for the workflow author to use past the exception if he or she see fit.

Cool? Let's move on this.

**Comment by shawnrice** [comment]
Also, I'm going to put something like this in the documentation for the PHP bundler:

```
// Here is some sample code as to how to instantiate the Bundler. If you care
// about keeping the script going, then you should use similiar code as below 
// in order to catch an exception if the user refuses to install the Bundler.

// If you want to make the workflow function without the bundler (after 
// refusal to install), then you should set a flag when the initial exception 
// is thrown and save it somewhere in the settings so that the user isn't 
// prompted to install the bundler each time the workflow is called.
if ( $settings[ 'alfredBundler' ] != 'noBundler' ) :

require_once( 'alfred.bundler.php' );

try {
    // Create the AlfredBundler object. If the Bundler is not installed, then
    // an installation of the Bundler will be attempted here.
    $b = new AlfredBundler;
} catch (Exception $e) {
    // If the user cancels the installation, then an exception will be thrown
    // so, catch it here and do what you will.

    // Also, an environmental flag will be set: 
    // $_ENV[ 'ALFRED_BUNDLER_INSTALL_REFUSED' ] = TRUE;
    // So, if you need to use that later when calling bundler functions, 
    // it's there.

    // Do exception handling
    echo 'Caught exception: ',  $e->getMessage(), "\n";
}

endif;
```

**Comment by shawnrice** [comment]
One last thing to note:

When the user presses "Cancel" there is no text returned. So:

```
button returned:Proceed
button returned:More Info
<blank>
```

I'm handling the reception in an `if/elseif/else` with the "Cancel" being the else. If that helps.

**Comment by shawnrice** [comment]
Python is the only one left, and since this issue is recorded in #49, I'm closing.
